app/assets/javascripts/map.js

```diff
@@ -125,25 +207,210 @@ class Map {
+    if (this.saltBagsQuantity > 0) {
+      this.saltBagsQuantity -= 1;
+      this.updateSaltBagsTotals();
+    }
+  }
+  
+  updateSaltBagsTotals() {
+    this.saltBagsDue = this.saltBagsQuantity * this.saltBagPrice;
+    this.updateTotalSaltBags();
+    this.updateGrandTotal();
+  }
+  
+  updateTotalSaltBags() {
+    this.updateAmount('numberOfBags', this.saltBagsQuantity, 0);
+    Dom.showNode(document.getElementById("summarySaltBags"));
+    this.updateAmount('summaryNumberOfBags', this.saltBagsQuantity, 0);
+    this.updateAmount('saltBagsDue', this.saltBagsDue, 2);
+  }
+
+  handleDoneAddOns() {
+    this.showNextSection('collapseAddOns', 'collapseContactAndPaymentInfo');
+    document.getElementById('checkMarkAddOns').classList.remove('hidden');
+
+  }
+
+  updateGrandTotal() {
+    let grandTotalBeforeDiscount = this.subTotal + this.serviceExpeditionCost + this.saltBagsDue;
+    this.updateDiscount(grandTotalBeforeDiscount);
+    this.grandTotal = Math.max(grandTotalBeforeDiscount, priceList.MIN_CHARGE) * this.discount;
+    if (this.grandTotal > priceList.MIN_CHARGE) {
+      Dom.hideNode(document.getElementById('minChargeNote'))
+    } else {
+      Dom.showNode(document.getElementById('minChargeNote'))
+    }
+    this.updateAmount('totalDue', this.grandTotal, 2)
+  }
+
+  updateDiscount(grandTotalBeforeDiscount) {
+    this.updateAmount('summaryDiscountAmount', this.discount * 100, 0)
+    this.updateAmount('totalDiscount', grandTotalBeforeDiscount * (- this.discount), 2)
+  }
+
+  onCreditCardInputChange(cardInputIsComplete) {
+    this.isCreditCardComplete = cardInputIsComplete;
+    this.enableSubmitButton();
+    Dom.hideNode(document.getElementById("contact_info_errors"))
+  }
+
+  enableSubmitButton() {
+    const requiredFieldsAreNotEmpty = ['quote_email', 'quote_phone_number'].every((id) => {
+      return document.getElementById(id).value !== "";
+    });
+
+    if (requiredFieldsAreNotEmpty && this.isCreditCardComplete) {
+      document.getElementById('submitPayment').classList.remove('disabled');
+    }
+  }
+
+  handlePaymentSubmit(event) {
+    event.preventDefault();
+    document.getElementById('submitPayment').classList.add('disabled');
+    this.stripe.getToken(this.handleTokenReceived);
+  }
+
+  handleTokenReceived(token) {
+    const form = document.getElementById('contact-info-form');
+    const formData = new FormData(form);
+    const coordinates = this.polygons.map(p => { 
+      return p.getPath().getArray().map(vertex => { return vertex.toJSON() })
+    });
+
+    const postData = {
+      quote: {
+        email: formData.get("quote[email]"),
+        phone_number: formData.get("quote[phone_number]"),
+        first_name: formData.get("quote[first_name]"),
+        last_name: formData.get("quote[last_name]"),
+        comments: formData.get("quote[comments]"),
+        address: this.geocodedAddress,
+        area: this.totalAreaInSqFt,
+        polygons_coordinates: JSON.stringify(coordinates),
```

Comment:
Why is this stringified? It's going to get _double_ stringified on L397, won't it?